File: src/test/java/net/scriptgate/masonry/demo/MasonryGui.java
package net.scriptgate.masonry.demo;

import net.scriptgate.common.Rectangle;
import net.scriptgate.engine.Application;
import net.scriptgate.engine.InputComponent;
import net.scriptgate.engine.Key;
import net.scriptgate.engine.Renderer;
import net.scriptgate.masonry.BasicMasonry;
import net.scriptgate.masonry.BrickListContainer;

import java.util.Random;

import static net.scriptgate.engine.Engine.HEIGHT;
import static net.scriptgate.engine.Engine.WIDTH;
import static net.scriptgate.masonry.Launcher.launch;

public class MasonryGui implements Application {

    public static void main(String[] args) {
        launch(new MasonryGui());
    }

    private BasicMasonry masonry;
    private BrickListContainer<ColorBrick> container;


    @Override
    public void initialize() {
        container = new BrickListContainer<>(WIDTH, HEIGHT);
        masonry = new BasicMasonry(container, 40);
    }

    @Override
    public void onKeyDown(Key key) {

    }

    @Override
    public void onTick(InputComponent inputComponent, double elapsedTime) {
        container.getBricks().stream().forEach(brick -> brick.update(elapsedTime));

        inputComponent.getPressedKeys().forEach(key -> {
            int width = (int) (Math.floor(2 * Math.random() + 1) * 40);
            int height = (int) (Math.floor(3 * Math.random() + 1) * 40);
            ColorBrick brick = new ColorBrick(width, height);
            switch (key.getKeyName()) {
                case "q":
                    container.addBrick(brick);
                    masonry.layout();
                    break;
                case "w":
                    container.addBrick(brick);
                    if (container.getBricks().isEmpty()) {
                        break;
                    }
                    int elementsToSkip = new Random().nextInt(container.getBricks().size());
                    container.getBricks().stream()
                            .skip(elementsToSkip)
                            .findFirst()
                            .ifPresent(b -> {
                                container.getBricks().remove(b);
                                masonry.layout();
                            });
                    break;
            }
        });
    }

    @Override
    public void onClick(int x, int y) {
        container.getBricks()
                .stream()
                .filter(brick -> {
                    Rectangle bounds = new Rectangle(brick.getX(), brick.getY(), brick.getWidth(), brick.getHeight());
                    return bounds.contains(x, y);
                })
                .findFirst().ifPresent(brick -> {
            container.removeBrick(brick);
            masonry.layout();
        });
    }

    @Override
    public void render(Renderer renderer) {
        renderer.setOpacity(0.5f);
        container.getBricks().forEach(brick -> {
            brick.render(renderer);
        });
    }
}

File: src/test/kotlin/net/scriptgate/masonry/demo/MasonryGui.kt
package net.scriptgate.masonry.demo

import net.scriptgate.common.Rectangle
import net.scriptgate.engine.Application
import net.scriptgate.engine.Engine.HEIGHT
import net.scriptgate.engine.Engine.WIDTH
import net.scriptgate.engine.InputComponent
import net.scriptgate.engine.Key
import net.scriptgate.engine.Renderer
import net.scriptgate.masonry.BasicMasonry
import net.scriptgate.masonry.BrickListContainer
import net.scriptgate.masonry.Launcher
import java.util.*

class MasonryGui : Application {

    private var count = 0

    private lateinit var masonry: BasicMasonry
    private lateinit var container: BrickListContainer<ColorBrick>


    override fun initialize() {
        container = BrickListContainer(WIDTH, HEIGHT)
        masonry = BasicMasonry(container, 40)
    }

    override fun onKeyDown(key: Key?) {

    }

    override fun onUpdate(ticks: Int, frames: Int) {
        println("ticks=$ticks frames=$frames count=${count++}")
        container.bricks.forEach { brick -> brick.update(ticks.toDouble()) }
    }

    override fun onTick(inputComponent: InputComponent, elapsedTime: Double) {
        inputComponent.pressedKeys.forEach { key ->
            val width = (Math.floor(2 * Math.random() + 1) * 40).toInt()
            val height = (Math.floor(3 * Math.random() + 1) * 40).toInt()
            val brick = ColorBrick(width, height)
            if (key.keyName != null) {

                when (key.keyName) {
                    "q" -> {

                        container.addBrick(brick)
                        masonry.layout()
                    }
                    "w" -> {
                        container.addBrick(brick)
                        val elementsToSkip = Random().nextInt(container.bricks.size)
                        container.bricks.stream()
                                .skip(elementsToSkip.toLong())
                                .findFirst()
                                .ifPresent { b ->
                                    container.bricks.minus(b)
                                    masonry.layout()
                                }
                    }
                }
            }
        }
    }

    override fun render(renderer: Renderer) {
        renderer.setOpacity(0.5f)
        container.bricks.forEach { brick: ColorBrick -> brick.render(renderer) }
        renderer.drawText(100,100, "Count: $count")
    }

    override fun onClick(x: Int, y: Int) {

        container.bricks
                .stream()
                .filter { brick ->
                    val bounds = Rectangle(brick.x, brick.y, brick.width, brick.height)
                    bounds.contains(x, y)
                }
                .findFirst().ifPresent { brick ->
                    container.removeBrick(brick)
                    masonry.layout()
                }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            Launcher.launch(MasonryGui())
        }
    }
}
